src/components/SigningModals/TransactionStepper/TransactionStepper.js

```diff
@@ -0,0 +1,247 @@
+        if (stepperStage !== stepsCount && canPerformUpdate.current) {
+          setStepperStage(stepperStage + 1)
+        }
+
+        if (stepperStage === stepsCount && canPerformUpdate.current) {
+          onComplete()
+        }
+      },
+    })
+  }, [
+    steps,
+    stepperStage,
+    updateStepStatus,
+    updateHash,
+    onComplete,
+    stepsCount,
+  ])
+
+  useEffect(handleSign, [stepperStage])
+
+  // Prevent async state updates after unmount
+  useEffect(() => {
+    return () => {
+      canPerformUpdate.current = false
+    }
+  }, [])
+
+  return (
+    <div className={className}>
+      <div
+        ref={outerBoundsRef}
+        css={`
+          display: flex;
+          justify-content: center;
+        `}
+      >
+        <ul
+          ref={innerBoundsRef}
+          css={`
+            padding: 0;
+            display: flex;
+            flex-direction: ${layout === 'single' ? 'column' : 'row'};
+            visibility: ${measuring ? 'hidden' : 'visible'};
+          `}
+        >
+          {layout === 'single' && (
+            <React.Fragment>
+              <p
+                css={`
+                  text-align: center;
+                  color: ${theme.contentSecondary};
+                `}
+              >
+                {stepperStage + 1} out of {steps.length} transactions
+              </p>
+
+              <div
+                css={`
+                  position: relative;
+                `}
+              >
+                <Transition
+                  config={springs.smooth}
+                  delay={300}
+                  items={stepperStage}
+                  immediate={firstStart}
+                  onStart={onStart}
+                  from={{
+                    opacity: 0,
+                    transform: `translate3d(${10 * GU}px, 0, 0)`,
+                  }}
+                  enter={{
+                    opacity: 1,
+                    transform: 'translate3d(0, 0, 0)',
+                  }}
+                  leave={{
+                    opacity: 0,
+                    transform: `translate3d(-${20 * GU}px, 0, 0)`,
+                  }}
+                  native
+                >
+                  {currentStage => animProps => (
+                    <AnimatedDiv
+                      style={{
+                        position:
+                          currentStage === stepperStage ? 'static' : 'absolute',
+                        ...animProps,
+                      }}
+                    >
+                      {renderStep(currentStage)}
+                    </AnimatedDiv>
+                  )}
+                </Transition>
+              </div>
+            </React.Fragment>
+          )}
+          {layout === 'multiple' && renderSteps()}
+        </ul>
+      </div>
+    </div>
+  )
+}
+
+TransactionStepper.defaultProps = {
+  onComplete: noop,
+}
+
+TransactionStepper.propTypes = {
+  steps: PropTypes.arrayOf(PropTypes.array),
```

Comment:
This should probably be marked as required—and should we also add a custom shape?

Reply:
I can't easily do this without adding a custom validator afaik. Though I might change this when I update the api to support custom descriptions.